src/pgwire/src/protocol.rs

```diff
@@ -943,8 +955,10 @@ where
         fetch_portal_name: Option<String>,
         timeout: ExecuteTimeout,
     ) -> BoxFuture<'_, Result<State, io::Error>> {
+        let execute_started = Instant::now();
```

Comment:
I think this should be move to just above the call to `adapter_client.execute`, identical to `one_query`. The postgres protocol works on portals (similar to network ports). Usually a pgwire client will want to execute a statement and get all result rows. But the protocol supports getting only some result rows and then coming back later on and getting more rows from that portal. In the more rows case, there's no statement being executed and we don't want to measure time to first row. Thus, I'd argue we should only report this metric in the `NotStarted` case.

I think it'd encourage correctness if there's a convenient way to merge this metric right into `adapter_client.execute` so that it'd work equally transparently for pgwire simple, pgwire extended, http and ws. Today each of these has to figure out when to start and stop measuring, and http and ws will require more code to measure this. Doing this might be kinda hard because of how the rows ExecuteResponse is structured (a future), so you'd have to wrap that future in a thing that could do all of this correctly. A bit tricky, but I think worth it. It'd certainly simplify a lot of the code here in this file (probably remove most or all of it).

Reply:
Most recent diff should have something like this

Reply:
Kinda frustrating, but when I tried to add the metric to sql.rs I wasn't able to use it as is... In https://github.com/MaterializeInc/materialize/blob/c8c1ec783ef790d1029f5ebda28d2851c529394a/src/environmentd/src/http/sql.rs#L1041-L1053 we don't create an UnboundedReceiver like we do in the postgres protocol. I'm not sure what to do. I played around with adding a RecordFirstRowFuture that does the same logic and wraps a RowsFuture and returns a RowsFuture. This works but then we have a fair bit of code that covers 2-3 usecases.